DSharpPlus/AsyncEvents/AsyncEvent`2.cs

```diff
@@ -0,0 +1,113 @@
+    /// </summary>
+    /// <typeparam name="TSender">The type of the object to dispatch this event.</typeparam>
+    /// <typeparam name="TArgs">The type of the argument object for this event.</typeparam>
+    public sealed class AsyncEvent<TSender, TArgs> : AsyncEvent
+        where TArgs : AsyncEventArgs
+    {
+        // we should look into not using a primitive lock here, but it's not a huge issue
+        private readonly object _lock = new();
+        private List<AsyncEventHandler<TSender, TArgs>> _handlers;
+        private readonly AsyncEventExceptionHandler<TSender, TArgs> _exceptionHandler;
+
+        public AsyncEvent(string name, AsyncEventExceptionHandler<TSender, TArgs> exceptionHandler)
+            : base(name)
+        {
+            this._handlers = new();
+            this._exceptionHandler = exceptionHandler;
+        }
+
+        /// <summary>
+        /// Registers a new handler for this event.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">Thrown if the specified handler was null.</exception>
+        public void Register(AsyncEventHandler<TSender, TArgs> handler)
+        {
+            if (handler is null)
+                throw new ArgumentNullException(nameof(handler));
+
+            lock (this._lock)
+                this._handlers.Add(handler);
+        }
+
+        /// <summary>
+        /// Unregisters a specific handler from this event.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">Thrown if the specified handler was null.</exception>
+        public void Unregister(AsyncEventHandler<TSender, TArgs> handler)
+        {
+            if (handler is null)
+                throw new ArgumentNullException(nameof(handler));
+
+            lock (this._lock)
+                this._handlers.Remove(handler);
+        }
+
+        /// <summary>
+        /// Unregisters all handlers from this event.
+        /// </summary>
+        public void UnregisterAll()
+            => this._handlers = new();
+
+        /// <summary>
+        /// Raises this event, invoking all registered handlers in parallel.
+        /// </summary>
+        /// <param name="sender">The instance that dispatched this event.</param>
+        /// <param name="args">The arguments passed to this event.</param>
+        public Task InvokeAsync(TSender sender, TArgs args)
+        {
+            if (this._handlers.Count == 0)
+                return Task.CompletedTask;
+
+            Parallel.ForEach
+            (
+                this._handlers,
+                handler =>
+                {
+                    try
+                    {
+                        handler.Invoke(sender, args);
+                    }
+                    catch (Exception ex)
+                    {
+                        this._exceptionHandler.Invoke(this, ex, handler, sender, args);
+                    }
+                }
+            );
```

Comment:
Correct me if I'm wrong, but if the handler returns a task (`public delegate Task`), why is `Parallel.ForEach` used in favor of `Task.WhenAll`? (**N.B:** The task isn't even awaited, so the handler would never catch the error?)

Reply:
another good call

Reply:
implemented